source/features/batch-mark-files-as-viewed.tsx

```diff
@@ -18,18 +18,26 @@ function isChecked(file: HTMLElement): boolean {
 }
 
 function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
-	if (!event.shiftKey || !previousFile) {
+	if (!event.shiftKey) {
+		return;
+	}
+
+	const isFirstToggle = previousFile === undefined;
+
+	previousFile ??= select('.js-file'); // #5484
+	if (!previousFile?.isConnected) {
 		return;
 	}
 
 	event.preventDefault();
 	event.stopImmediatePropagation();
 
-	const previousFileState = isChecked(previousFile);
 	const thisFile = event.delegateTarget.closest('.js-file')!;
 	const files = select.all('.js-file');
+	const previousFileState = isFirstToggle ? !isChecked(thisFile) : isChecked(previousFile);
```

Comment:
Our logic was a little silly and we're carrying it over here. Let me explain:

- if I click an **unchecked toggle**, I want it to become **checked**
- if I alt-click an **unchecked toggle**, I want _all_ the (related) toggles to become **checked**

What we were doing on line 28 was discarding the current toggle and just using the _previous_’ toggle state. Example:

1. Start with:

	- [ ] a
	- [ ] b
	- [x] c

2. Click on `a`
3. Alt-click on `c`

You'll see this weird behavior (on `main`, not on this PR):


https://user-images.githubusercontent.com/1402241/163171226-9b00a756-c925-4f11-a067-f8ab6e9092ad.mov



This condition should probably be `isChecked(thisFile)` (possibly negated, depending on whether the listener is running before or after the state is changed by the native click)

Reply:
@fregante

Thanks for your detailed descriptions :)
I organized my thoughts here.

> This condition should probably be isChecked(thisFile)

I agree with this point.
I think maybe the below code using `isChecked(thisFile)` instead of `previousFileState` should work.

```ts
function batchToggle(event: delegate.Event<MouseEvent, HTMLFormElement>): void {
  if (!event.shiftKey) {
    return;
  }

  previousFile ??= select('.js-file'); // #5484
  if (!previousFile?.isConnected) {
    return;
  }

  event.preventDefault();
  event.stopImmediatePropagation();

  const thisFile = event.delegateTarget.closest('.js-file')!;
  const files = select.all('.js-file');

  const selections = [
    files.indexOf(previousFile),
    files.indexOf(thisFile),
  ].sort((a, b) => a - b);

  selections[1] += 1;

  const selectedFiles = files.slice(...selections);

  for (const file of selectedFiles) {
    if (isChecked(file) === isChecked(thisFile)) {
      select('.js-reviewed-checkbox', file)!.click();
    }
  }
}
```

> it doesn't break/change the regular behavior.

I think the changed code seems to affect only `isFirstToggle` is true.
So, it works as existing code (as expected) when `isFirstToggle` is false.

> this new behavior works

Yes, new behavior works as expected.

Reply:
> breaks existing behaviors

If it works correctly, that's ok. I just meant that the behavior I described is unrelated to this PR and doesn't _have to_ be fixed here. If you want to do that, feel free to.

Reply:
I kinda wish this code was testable. If you decide to fix both, can you extract this logic into a file under `source/helpers`?

Maybe something like this, which should not consider `isChecked` at all:

```ts
/** Get the items between previous and current, both ends included. If `previous` is missing, start from 0 */
function getItemsBetween<T>(items: T[], previous: T | undefined, current: T): T[] {

}
```

I think once you extract this, the rest of the function should be nearly unchanged. There's already `isChecked` in the last loop that takes care of the skipping unrelated items.

The core feature has 3 parts:

1. Determining the final state: set it to `thisFile`, unconditionally
2. Getting the items between previous and current
3. Setting all the items to the final state we already know, skipping those that are already that

Reply:
👆 `getItemsBetween` is the only logic that needs to be tested, making sure that it works even when previous is undefined or if its index comes after `current` (that's already taken care by `sort`)

Reply:
> If you decide to fix both, can you extract this logic into a file under source/helpers?

I'd like to fix the other part if possible (in another PR).
I just refactored the code using `getItemsBetween`.

Reply:
> I'd like to fix the other part if possible (in another PR).

I think that's as easy as using the current file’s state instead of the `previous` file. We just need to ensure that `isChecked` works on the current file. By "works" I mean:

- if I'm **checking** the file, then it should return `true` immediately